fix failing dryrun e2e job

some of our recent changes broke the dryrun e2e

https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-dryrun-latest

tip: to see what commits was it one can click a red box and drag to a green box. i am currently on mobile and can't do that.

we should try to fix this before xmas and not having any failing tests during the holidays...preferably since folks are not around.

@pacoxu @SataQiu @chendave

[Here](https://github.com/kubernetes/kubernetes/commit/e4353d3ec21e13f05a1986b1eaf3c18fa8f224cd) is my attempt at a local repo of the issue.

```
=== RUN   TestDiscoveryServerVersion
[dryrun] Would perform action GET on resource "version" in API group "core/"
    dryrunclient_test.go:39: Get ServerVersion failed.: can't cast get reactor event action object of type testing.ActionImpl to GetAction interface
--- FAIL: TestDiscoveryServerVersion (6.96s)
```

So the cast fails because `action` is of type `testing.ActionImpl` (created by fake discovery [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/fake/discovery.go#L141)) which can't be cast to `core.GetAction`.

It's still not clear to me how this cast could have worked in the past. As I suggested above, maybe the cast failed in the past but prior to https://github.com/kubernetes/kubernetes/pull/114291 the resulting error was ignored?

--- a/cmd/kubeadm/app/util/apiclient/dryrunclient.go
+++ b/cmd/kubeadm/app/util/apiclient/dryrunclient.go
@@ -119,8 +119,16 @@ func NewDryRunClientWithOpts(opts DryRunClientOptions) clientset.Interface {
 			Reaction: func(action core.Action) (bool, runtime.Object, error) {
 				getAction, ok := action.(core.GetAction)
 				if !ok {
-					// something's wrong, we can't handle this event
-					return true, nil, errors.New("can't cast get reactor event action object to GetAction interface")
+					// If the GetAction cast fails, this could be an ActionImpl with a "get" verb.
+					// Such actions could be invoked from any of the fake discovery calls, such as ServerVersion().
+					// Attempt the cast to ActionImpl and construct a GetActionImpl from it.
+					actionImpl, ok := action.(core.ActionImpl)
+					if ok {
+						getAction = core.GetActionImpl{ActionImpl: actionImpl}
+					} else {
+						// something's wrong, we can't handle this event
+						return true, nil, errors.New("can't cast get reactor event action object to GetAction interface")
+					}
 				}
 				handled, obj, err := opts.Getter.HandleGetAction(getAction)
 

--- a/cmd/kubeadm/app/util/apiclient/dryrunclient_test.go
+++ b/cmd/kubeadm/app/util/apiclient/dryrunclient_test.go
@@ -18,12 +18,15 @@ package apiclient
 
 import (
 	"bytes"
+	"io"
 	"testing"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	rbac "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	pkgversion "k8s.io/apimachinery/pkg/version"
+	fakediscovery "k8s.io/client-go/discovery/fake"
 	core "k8s.io/client-go/testing"
 )
 
@@ -107,5 +110,26 @@ func TestLogDryRunAction(t *testing.T) {
 				)
 			}
 		})
+	}
+}
+
+func TestDiscoveryServerVersion(t *testing.T) {
+	dryRunGetter := &InitDryRunGetter{
+		controlPlaneName: "controlPlane",
+		serviceSubnet:    "serviceSubnet",
+	}
+	c := NewDryRunClient(dryRunGetter, io.Discard)
+	fakeclientDiscovery, ok := c.Discovery().(*fakediscovery.FakeDiscovery)
+	if !ok {
+		t.Fatal("could not obtain FakeDiscovery from dry run client")
+	}
+	const gitVersion = "foo"
+	fakeclientDiscovery.FakedServerVersion = &pkgversion.Info{GitVersion: gitVersion}
+	ver, err := c.Discovery().ServerVersion()
+	if err != nil {
+		t.Fatalf("Get ServerVersion failed.: %v", err)
+	}
+	if ver.GitVersion != gitVersion {
+		t.Fatalf("GitVersion did not match, expected %s, got %s", gitVersion, ver.GitVersion)
 	}
 }

